include/rcutils/filesystem.h

```diff
@@ -185,19 +192,46 @@ bool
 rcutils_mkdir(const char * abs_path);
 
 /// Calculate the size of the specified directory.
-/*
+/**
  * Calculates the size of a directory by summarizing the file size of all files.
  * \note This operation is not recursive.
  * \param[in] directory_path The directory path to calculate the size of.
  * \param[in] allocator Allocator being used for internal file path composition.
- * \return The size of the directory in bytes.
+ * \return The size of the directory in bytes on success.
+ * \return -1 on failure.
  */
 RCUTILS_PUBLIC
-size_t
+ssize_t
 rcutils_calculate_directory_size(const char * directory_path, rcutils_allocator_t allocator);
 
+/// Calculate the size of the specified directory with recursion.
+/**
+ * Calculates the size of a directory and subdirectory by summarizing the file size of all files.
+ * If necessary, you can specify the maximum directory depth to recurse into.
+ * Depth definition as below.
+ * \code
+ * directory_path  <= depth 1
+ *    |- subdirectory <= depth 2
+ *            |- subdirectory <= depth 3
+ *                    ...
+ * \endcode
+ *
+ * \note This operation doesn't follow the symbolic link that points to directory.
```

Comment:
Can you clarify the situation with symlinks a bit, and maybe add tests for it?  In particular:

1.  What happens if there is a symlink to a file?  Does the size of that file get added to the overall returned size?
1.  What happens if there is a symlink to a directory?  Do we recurse into that directory and add the sizes?